# Autograde a students' solution

```
submitted/{student_id}/{assignment_id}/{notebook_id}.ipynb
```

After running `nbgrader autograde`, the autograded version of the notebooks will be:

```
autograded/{student_id}/{assignment_id}/{notebook_id}.ipynb
```

## Example

In the following example, we have an assignment with two notebooks. There are two submissions of the assignment:

Submission 1:

* [submitted/Bitdiddle/Problem Set 1/Problem 1.ipynb](submitted/Bitdiddle/Problem Set 1/Problem 1.ipynb)
* [submitted/Bitdiddle/Problem Set 1/Problem 2.ipynb](submitted/Bitdiddle/Problem Set 1/Problem 2.ipynb)

Submission 2:

* [submitted/Hacker/Problem Set 1/Problem 1.ipynb](submitted/Hacker/Problem Set 1/Problem 1.ipynb)
* [submitted/Hacker/Problem Set 1/Problem 2.ipynb](submitted/Hacker/Problem Set 1/Problem 2.ipynb)

Before we can actually start grading, we need to actually record who the students are. We can do this using the API provided by nbgrader, which provides access to a database to store information about students and their grades:

In [1]:
# create a connection to the db using the nbgrader API
from nbgrader.api import Gradebook
gb = Gradebook("sqlite:///gradebook.db")

# add some students to the database
gb.add_student("Bitdiddle", first_name="Ben", last_name="Bitdiddle")
gb.add_student("Hacker", first_name="Alyssa", last_name="Hacker")
gb.add_student("Reasoner", first_name="Louis", last_name="Reasoner")

Student<Reasoner>

Once the database has been set up with the students, we can run the autograder:

In [2]:
%%bash

nbgrader autograde "Problem Set 1"

[AutogradeApp | INFO] Linking submitted/Bitdiddle/Problem Set 1/timestamp.txt -> autograded/Bitdiddle/Problem Set 1/timestamp.txt
[AutogradeApp | INFO] Linking submitted/Bitdiddle/Problem Set 1/jupyter.png -> autograded/Bitdiddle/Problem Set 1/jupyter.png
[AutogradeApp | INFO] SubmittedAssignment<Problem Set 1 for Bitdiddle> submitted at 2015-02-02 14:58:23.948203
[AutogradeApp | WARNING] SubmittedAssignment<Problem Set 1 for Bitdiddle> is 86303.948203 seconds late
[AutogradeApp | INFO] Overwriting files with master versions from the source directory
[AutogradeApp | INFO] Linking source/./Problem Set 1/jupyter.png -> autograded/Bitdiddle/Problem Set 1/jupyter.png
[AutogradeApp | INFO] Sanitizing submitted/Bitdiddle/Problem Set 1/Problem 2.ipynb
[AutogradeApp | INFO] Converting notebook submitted/Bitdiddle/Problem Set 1/Problem 2.ipynb to notebook
[AutogradeApp | INFO] Writing 2556 bytes to autograded/Bitdiddle/Problem Set 1/Problem 2.ipynb
[AutogradeApp | INFO] Autograding autograded/B

When grading the submission for `Bitdiddle`, you'll see some warnings that look like "Checksum for grade cell correct_squares has changed!". What's happening here is that nbgrader has recorded what the *original* contents of the grade cell `correct_squares` (when `nbgrader assign` was run), and is checking the submitted version against this original version. It has found that the submitted version changed (perhaps this student tried to cheat by commenting out the failing tests), and has therefore overwritten the submitted version of the tests with the original version of the tests.

You may also notice that there is a note saying "Problem Set 1 for Bitdiddle is 86303.948203 seconds late". What is happening here is that nbgrader is detecting a file in Bitdiddle's submission called `timestamp.txt`, reading in that timestamp, and saving it into the database. From there, it can compare the timestamp to the duedate of the problem set, and compute whether the submission is at all late.

Once the autograding is complete, there will be new directories for the autograded versions of the submissions:

Autograded submission 1:

* [autograded/Bitdiddle/Problem Set 1/Problem 1.ipynb](autograded/Bitdiddle/Problem Set 1/Problem 1.ipynb)
* [autograded/Bitdiddle/Problem Set 1/Problem 2.ipynb](autograded/Bitdiddle/Problem Set 1/Problem 2.ipynb)

Autograded submission 2:

* [autograded/Hacker/Problem Set 1/Problem 1.ipynb](autograded/Hacker/Problem Set 1/Problem 1.ipynb)
* [autograded/Hacker/Problem Set 1/Problem 2.ipynb](autograded/Hacker/Problem Set 1/Problem 2.ipynb)